File: FBControlCore/Configuration/FBProcessSpawnConfiguration.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBProcessLaunchConfiguration.h"

#import <FBControlCore/FBControlCore.h>

@implementation FBProcessSpawnConfiguration

- (instancetype)initWithLaunchPath:(NSString *)launchPath arguments:(NSArray<NSString *> *)arguments environment:(NSDictionary<NSString *, NSString *> *)environment io:(FBProcessIO *)io mode:(FBProcessSpawnMode)mode;
{
  self = [super initWithArguments:arguments environment:environment io:io];
  if (!self) {
    return nil;
  }

  _launchPath = launchPath;
  _mode = mode;

  return self;
}

- (NSString *)processName
{
  return self.launchPath.lastPathComponent;
}

#pragma mark NSObject

- (NSUInteger)hash
{
  return [super hash] | self.launchPath.hash | self.mode;
}

- (BOOL)isEqual:(FBProcessSpawnConfiguration *)object
{
  if (![object isKindOfClass:self.class]) {
    return NO;
  }
  return [self.launchPath isEqual:object.launchPath] &&
         [self.arguments isEqual:object.arguments] &&
         [self.environment isEqual:object.environment] &&
         self.mode == object.mode;
}

- (NSString *)description
{
  return [NSString stringWithFormat:
    @"Process Launch %@ | Arguments %@ | Environment %@ | Output %@",
    self.launchPath,
    [FBCollectionInformation oneLineDescriptionFromArray:self.arguments],
    [FBCollectionInformation oneLineDescriptionFromDictionary:self.environment],
    self.io
  ];
}

@end

File: FBControlCore/Configuration/FBProcessSpawnConfiguration.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation

@objc
public enum ProcessSpawnMode: UInt {
  case `default` = 0
  case posixSpawn = 1
  case launchd = 2
}

@objc
public final class FBProcessSpawnConfiguration: ProcessLaunchConfiguration {

  @objc public let launchPath: String
  @objc public let mode: ProcessSpawnMode

  @objc public var processName: String {
    (launchPath as NSString).lastPathComponent
  }

  @objc
  public init(launchPath: String, arguments: [String], environment: [String: String], io: FBProcessIO<AnyObject, AnyObject, AnyObject>, mode: ProcessSpawnMode) {
    self.launchPath = launchPath
    self.mode = mode
    super.init(arguments: arguments, environment: environment, io: io)
  }

  public override var hash: Int {
    super.hash | (launchPath as NSString).hash | Int(mode.rawValue)
  }

  public override func isEqual(_ object: Any?) -> Bool {
    guard let other = object as? FBProcessSpawnConfiguration,
      other.isKind(of: type(of: self))
    else {
      return false
    }
    return launchPath == other.launchPath
      && (arguments as NSArray).isEqual(to: other.arguments)
      && (environment as NSDictionary).isEqual(to: other.environment as NSDictionary)
      && mode == other.mode
  }

  public override var description: String {
    "Process Launch \(launchPath) | Arguments \(CollectionInformation.oneLineDescription(from: arguments)) | Environment \(CollectionInformation.oneLineDescription(from: environment)) | Output \(io)"
  }
}
